ibis/expr/format.py

```diff
@@ -138,31 +138,33 @@ def get_result(self):
     def _memoize_tables(self):
         table_memo_ops = (ops.Aggregation, ops.Selection,
                           ops.SelfReference)
-
-        def walk(expr):
-            if id(expr) in self.memo.visit_memo:
-                return
-
-            op = expr.op()
-
-            def visit(arg):
-                if isinstance(arg, list):
-                    [visit(x) for x in arg]
-                elif isinstance(arg, ir.Expr):
-                    walk(arg)
-
-            if isinstance(op, ops.PhysicalTable):
-                self.memo.observe(expr, self._format_table)
-            elif isinstance(op, ir.Node):
-                visit(op.args)
-                if isinstance(op, table_memo_ops):
-                    self.memo.observe(expr, self._format_node)
-            elif isinstance(op, ops.TableNode) and op.has_schema():
-                self.memo.observe(expr, self._format_table)
-
-            self.memo.visit_memo.add(id(expr))
-
-        walk(self.expr)
+        if id(self.expr) in self.memo.visit_memo:
+            return
+
+        stack = [self.expr]
+        seen = set()
+        memo = self.memo
+
+        while stack:
+            e = stack.pop()
+            op = e.op()
+
+            if op not in seen:
+                seen.add(op)
+
+                if isinstance(op, ops.PhysicalTable):
+                    memo.observe(e, self._format_table)
+                elif isinstance(op, ir.Node):
+                    stack.extend(
+                        arg for arg in reversed(op.args)
+                        if isinstance(arg, ir.Expr)
+                    )
+                    if isinstance(op, table_memo_ops):
+                        memo.observe(e, self._format_node)
+                elif isinstance(op, ops.TableNode) and op.has_schema():
+                    memo.observe(e, self._format_table)
+
+                memo.visit_memo.add(id(e))
```

Comment:
should we dig up the past issues where there were performance problems in complicated expressions and add some asv benchmarks? otherwise we might introduce regressions in these refactorings